File: abc389/c/c.cpp
#include <iostream>
#include <deque>

using namespace std;

int main(){
    int Q;
    cin >> Q;
    long long p = 0;
    deque<long long> d;
    d.push_back(0);
    while(Q--){
        int q, l, k;
        cin >> q;
        if(q == 1){
            cin >> l;
            if(d.empty()) d.push_back(l);
            else d.push_back(l+d.back());
        }else if(q == 2){
            d.pop_front();
            if(!d.empty())p = d.front();
        }else{
            cin >> k;
            cout << d[k-1]-p << endl;
        }
    }
}

File: abc389/c/c.go
package main

import (
	"bufio"
	"fmt"
	"os"

	"github.com/liyue201/gostl/ds/deque"
)

func main() {
	r := bufio.NewReader(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	var Q int
	fmt.Fscan(r, &Q)
	dq := deque.New[int]()
	p := 0
	dq.PushBack(p)
	for i := 0; i < Q; i++ {
		var q, l, k int
		fmt.Fscan(r, &q)
		if q == 1 {
			fmt.Fscan(r, &l)
			if dq.Empty() {
				dq.PushBack(l)
			} else {
				dq.PushBack(l + dq.Back())
			}
		} else if q == 2 {
			dq.PopFront()
			if !dq.Empty() {
				p = dq.Front()
			}
		} else {
			fmt.Fscan(r, &k)
			fmt.Fprintln(w, dq.At(k-1)-p)
		}
	}
}
